common/data_refinery_common/models/models.py

```diff
@@ -698,6 +698,48 @@ def delete_local_file(self):
         self.is_downloaded = False
         self.save()
 
+    def needs_downloading(self, pipeline_applied=None) -> bool:
+        """Determine if a file needs to be downloaded.
+
+        This is true if the file has already been downloaded and lost
+        without getting processed.
+        """
+        # If the file is downloaded and the file actually exists on disk,
+        # then it doens't need to be downloaded.
+        if self.is_downloaded \
+           and self.absolute_file_path \
+           and os.path.exists(self.absolute_file_path):
+            return False
+
+        unstarted_downloader_jobs = self.downloader_jobs.filter(
+            start_time__isnull=True,
+            success__isnull=True,
+            retried=False
+        )
+
+        # If the file has a downloader job that hasn't even started yet,
+        # then it doesn't need another.
+        if unstarted_downloader_jobs.count() > 0:
+            return False
+
+        # Transcriptome files are used by two jobs, one for long and
+        # one for short. So one of them could have completed
+        # successfully before the file disappeared, therefore check
+        # the pipeline_applied to make sure we're looking at the same
+        # index_length.
+        if pipeline_applied:
+            successful_processor_jobs = self.processor_jobs.filter(
+                success=True,
+                pipeline_applied=pipeline_applied
+            )
+        else:
+            successful_processor_jobs = self.processor_jobs.filter(success=True)
+
+        # Finally, if there is a successful processor job, then the file
+        # has been processed and doesn't need to be processed again.
+        # XXX: if we ever want to reprocess everything, this will be problematic.
```

Comment:
Can we come up with a better solution than this, I'd rather not be introducing `XXX`es at this stage of development if they don't have a plan.

Reply:
I'm just gonna take this out then. I think that if/when we want to reprocess everything then we're gonna have to do some planning and come up with a way to keep the first and subsequent runs separate. Maybe just like a field denoting which version jobs were run on.